src/vmm/src/vmm_config/net.rs

```diff
@@ -116,140 +104,94 @@ impl Display for NetworkInterfaceError {
-    ) -> result::Result<(), NetworkInterfaceError> {
-        match self
-            .if_list
-            .iter()
-            .position(|netif_from_list| netif_from_list.iface_id == netif_config.iface_id)
-        {
-            Some(index) => self.update(index, netif_config),
-            None => self.create(netif_config),
-        }
-    }
-
-    fn get_index_of_mac(&self, mac: MacAddr) -> Option<usize> {
-        self.if_list
-            .iter()
-            .position(|netif| netif.guest_mac == Some(mac))
-    }
-
-    fn get_index_of_dev_name(&self, host_dev_name: &str) -> Option<usize> {
-        self.if_list
-            .iter()
-            .position(|netif| netif.host_dev_name == host_dev_name)
-    }
-
-    fn validate_update(
-        &self,
-        index: usize,
-        new_config: &NetworkInterfaceConfig,
-    ) -> result::Result<(), NetworkInterfaceError> {
-        // Check that the mac address is unique. In order to do so, we search for the
-        // network interface that has the same mac address as the one specified in new_config.
-        // If the same mac is used in another network interface config, return error.
-        if new_config.guest_mac.is_some() {
-            let mac_index = self.get_index_of_mac(new_config.guest_mac.unwrap());
-            if mac_index.is_some() && mac_index.unwrap() != index {
-                return Err(NetworkInterfaceError::GuestMacAddressInUse(
-                    new_config.guest_mac.unwrap().to_string(),
-                ));
-            }
-        }
-        // Check that the host_dev_name is unique.
-        let dev_name_index = self.get_index_of_dev_name(&new_config.host_dev_name);
-        if dev_name_index.is_some() && dev_name_index.unwrap() != index {
-            return Err(NetworkInterfaceError::HostDeviceNameInUse(
-                new_config.host_dev_name.clone(),
-            ));
-        }
-
-        Ok(())
-    }
-
-    fn update(
-        &mut self,
-        index: usize,
-        updated_netif_config: NetworkInterfaceConfig,
-    ) -> result::Result<(), NetworkInterfaceError> {
-        self.validate_update(index, &updated_netif_config)?;
-        self.if_list[index] = updated_netif_config;
-
-        // Check that the tap can be opened.
-        self.if_list[index].open_tap().map(|_| ())
-    }
-
-    fn validate_create(
-        &self,
-        new_config: &NetworkInterfaceConfig,
-    ) -> result::Result<(), NetworkInterfaceError> {
-        // Check that there is no other interface in the list that has the same mac.
-        if new_config.guest_mac.is_some()
-            && self
-                .get_index_of_mac(new_config.guest_mac.unwrap())
-                .is_some()
-        {
+    /// Inserts `netif_config` in the network interfaces list.
+    /// If an entry with the same id already exists, it will overwrite it.
+    pub fn insert(&mut self, netif_config: NetworkInterfaceConfig) -> Result<()> {
+        let no_mac_conflict = |net: &Arc<Mutex<Net>>| {
```

Comment:
Why not validate this at the upper layers ?

Reply:
There are no upper layers which _understand_ the meaning of the config fields. I believe this is the first (and ideally only) layer that should know how to interpret the config data.